### PR TITLE
[InstCombine] enable more factorization in SimplifyUsingDistributiveLaws

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
This change is from [D136623](https://reviews.llvm.org/D136623) but rebased on top of `main`. I'm going to import the relevant review comments from the original Phabricator revision and continue where we left off.

See #57255.

CC: @hiraditya

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 1]
```diff
@@ -997,11 +1009,33 @@ Value *InstCombinerImpl::foldUsingDistributiveLaws(BinaryOperator &I) {
...
     return R;
 
+  if (Op0)
+    LHSOpcode = getBinOpsForFactorization(TopLevelOpcode, Op0, A, B);
+  if (Op1)
+    RHSOpcode = getBinOpsForFactorization(TopLevelOpcode, Op1, C, D);
+
+  // The instruction has the form "(A op' B) op' (C * D)". See if expanding it
+  // out to "(C * D) op' (A op' B)" results in simplifications.
+  if (Op0 && Op1 && RHSOpcode == Instruction::Mul && A == C &&
```
Comment (tetsuo-cpp) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1026]: > **spatel**: It seems odd for this function to specify the exact opcode. Everything else is generalized based on association and distribution laws. Is mul really the only opcode where this transform works?

> **Allen (Author)**: Thanks, I restrict to mul as some performance regression of logic combination tests

Comment (tetsuo-cpp) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1026]: If this is still a sticking point, I can investigate what regressions Allen was talking about and explain why exactly we can't generalise this.

**Review by nikic — COMMENTED** [review]
Just to get the obvious stuff out of the way:
 * Please update the patch description to describe what this patch is doing.
 * Add alive2 proofs.
 * It looks like you are adding two related but independent transforms here -- I would recommend splitting them into separate patches.
 * And yes, we should not hardcode mul opcodes -- if we do, there should be clear justification for doing that.

File: llvm/test/Transforms/InstCombine/add4.ll [hunk 2]
```diff
@@ -127,3 +129,184 @@ define i32 @not_match_overflow(i32 %x) {
...
+; CHECK-NEXT:    [[A1:%.*]] = add i32 [[A0]], [[S]]
+; CHECK-NEXT:    ret i32 [[A1]]
+;
+  %a0 = add i32 %x, %y
+  %s = shl i32 %x, 2
+  %a1 = add i32 %a0, %s ; the shl operand is the right operand
+  ret i32 %a1
+}
+
+; Negative test: y is not a const for (y + x) + (x << 2)
```
Comment (hiraditya) [inline llvm/test/Transforms/InstCombine/add4.ll:271]: nit: this comment is a copy of previous?

**Comment by goldsteinn** [comment]
> Just to get the obvious stuff out of the way:
> 
> * Please update the patch description to describe what this patch is doing.
> * Add alive2 proofs.
> * It looks like you are adding two related but independent transforms here -- I would recommend splitting them into separate patches.
> * And yes, we should not hardcode mul opcodes -- if we do, there should be clear justification for doing that.

Can you also split the tests commit so we can see the IR diff generated by this patch?

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 3]
```diff
@@ -984,6 +984,18 @@ Value *InstCombinerImpl::tryFactorizationFolds(BinaryOperator &I) {
               tryFactorization(I, SQ, Builder, RHSOpcode, LHS, Ident, C, D))
         return V;
 
+  // The instruction has the form "(A * B) op (C op D)".  Try to factorize
+  // common term for "(A * B) op C op D".
+  if (Op0 && Op1 && LHSOpcode == Instruction::Mul && isa<Constant>(D) &&
+      LHS->hasOneUse() && RHS->hasOneUse() && TopLevelOpcode == RHSOpcode &&
+      Instruction::isCommutative(RHSOpcode))
+    if (Value *Ident = getIdentityValue(LHSOpcode, C))
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:992]: Why are you taking identity op of `LHS` and then using in RHS? Looks like you are doing the following:
```(A * B) + (C + D) -> factorize((A * B) + (C + 1)) + D``` which doesn't seem correct.

**Comment by dtcxzyw** [comment]
@tetsuo-cpp Could you please rebase this patch on https://github.com/llvm/llvm-project/commit/af2d740d2d9f3ff7ce9c038c26134c921f0662ae?

**Comment by tetsuo-cpp** [comment]
Closing this PR as I don't have time to work on it. @gxyd will be looking into this.

**Pull request closed without merging**